File: FileExchangeSystem_Client.java
// Machine Project Group 10
// TAPIA, John Lorenzo N.
// ARGAMOSA, Daniel Cedric S.

package FileExchangeSystem_Client;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.File;
import java.io.IOException;
import java.net.Socket;
import java.net.InetSocketAddress;
import java.net.SocketTimeoutException;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class FileExchangeSystem_Client {
    // Function to display Help page - accessed with '/?' command
    public static void displayHelp() {
        System.out.println(""
                + "/?\t-\tDisplay information about all valid commands\n"
                + "/dir\t-\tDisplay a directory of the files in the File Exchange server\n"
                + "/exit\t-\tExit the program - disconnects from any connected File Exchange server\n"
                + "/get [filename]\t-\tDownload a file of name [filename] from the File Exchange server\n"
                + "/join [ip_address] [port]\t-\tConnect to the File Exchange server of IP address [ip_address] and port number [port]\n"
                + "/leave\t-\tDisconnect from the connected File Exchange server\n"
                + "/register [alias]\t-\tRegister to the connected File Exchange server using the handle [alias] (must be unique)\n"
                + "/rejoin\t-\tReconnect to the previously-connected File Exchange server\n"
                + "/store [filename]\t-\tUpload a file of name [filename] to the File Exchange server - [filename] must exist inside the client's directory\n");
    }
    
    // Function to check whether a string is a valid IPv4 address
    public static boolean isValidIPAddress(String ipAddress) {
        // Define the regex pattern for a valid IPv4 address
        String ipPattern = "^(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\\."
                        + "(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\\."
                        + "(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)\\."
                        + "(25[0-5]|2[0-4][0-9]|[0-1]?[0-9][0-9]?)$";

        // Compile the pattern
        Pattern pattern = Pattern.compile(ipPattern);

        // Match the given IP address with the pattern
        Matcher matcher = pattern.matcher(ipAddress);

        // Return true if it matches or if the string is "localhost", false otherwise
        return ipAddress.equals("localhost") || matcher.matches();
    }
    
    // Function to connect a socket to a server given IPv4 address and port number
    public static boolean connectSocket(Socket soc, String ipAddress, int port) {
        try {
            System.out.println("Connecting to server at " + ipAddress + ":" + port);
            
            // Connect socket
            soc.connect(new InetSocketAddress(ipAddress, port), 6000); // Timeout after 6000 ms (6 seconds)
            
            System.out.println("Connected to server at " + soc.getRemoteSocketAddress());
            return true;
        } catch (SocketTimeoutException e) {
            // Close client socket if connection timed out
            try {
                soc.close();
                System.err.println("ERROR: Connection timed out. Check server availability and try again.");
            } catch (IOException e1) {
                System.err.println("I/O ERROR: " + e1.getMessage());
            }
        } catch (IOException e) {
            if (e instanceof java.net.ConnectException) {
                // Close client socket if server is offline
                try {
                    soc.close();
                    System.err.println("ERROR: Could not connect to the server. Make sure the server is running and try again.");
                } catch (IOException e1) {
                    System.err.println("I/O ERROR: " + e1.getMessage());
                }
            } else {
                System.err.println("I/O ERROR: " + e.getMessage());
            }
        }
        
        return false;
    }
    
    // Main function
    public static void main(String[] args) {
        // Initial interface
        System.out.println("------------------------------");
        System.out.println("FILE EXCHANGE SYSTEM");
        System.out.println("------------------------------\n");
        System.out.println("Welcome!");
        System.out.println("Please connect to a server to start...\n");
        System.out.println("To get a list of all recognized commands, enter \"/?\"");
        
        // Declare variables
        String sServerAdd = "";
        String sClientAlias = "";
        int nServerPort = -1;
        boolean bContinue = true;
        Socket socEndpoint = new Socket();
        DataInputStream disInput = null;
        DataOutputStream dosOutput = null;
        
        // Close client socket before entering main program
        try {
            socEndpoint.close();
        } catch (IOException e) {
            System.err.println("I/O ERROR: " + e.getMessage());
        }
        
        /*
        -----------------------------------------
        CLIENT-SERVER MESSAGE GUIDE
        
        MESSAGE             MEANING
        D                   Get directory
        F                   Get file
        L                   Close connection
        R                   Register
        S                   Store file
        -----------------------------------------
        */

        // Main program loop
        do {
            System.out.print("\n> ");
            
            // Scan for input
            Scanner scInputLine = new Scanner(System.in);
            String sInputLine = scInputLine.nextLine();
            Scanner scInput = new Scanner(sInputLine);
            System.out.println();

            // Evaluate input
            String sCommand = scInput.next();
            switch (sCommand) {
                // '/?' command
                case "/?" -> displayHelp();
                
                // '/dir' command
                case "/dir" -> {
                    // Check if (1) the client is connected to a server, (2) if the client is registered to the server
                    if (socEndpoint.isClosed()) {
                        System.out.println("ERROR: You are not connected to any File Exchange server.");
                    } else if (sClientAlias.isEmpty()) {
                        System.out.println("ERROR: You are not yet registered to this File Exchange server.");
                    } else {
                        try {
                            // Send request to server for its file directory
                            dosOutput.writeUTF("D");
                            
                            // Get the first response of server, which for this case should be an integer
                            // indicating the amount of files in the server
                            int nFileAmount = disInput.readInt();
                            
                            // Evaluate first response
                            if (nFileAmount == 0) {
                                System.out.println("Server directory is currently empty.");
                            } else {
                                System.out.println("List of current files in the server:");
                                // Print each file name received from the server
                                // up until nFirstResponse, which should contain the number of files in the server
                                for (int i = 0; i < nFileAmount; i++) {
                                    String sResponse = disInput.readUTF();
                                    
                                    // Check if the server sent the fail code, otherwise print the file name
                                    if (sResponse.equals("*")) {
                                        System.out.println("ERROR: An error occurred in the server.");
                                        i = nFileAmount; // Terminate loop
                                    } else {
                                        System.out.println(sResponse);
                                    }
                                }
                            }
                        } catch (IOException e) {
                            System.err.println("I/O ERROR: " + e.getMessage());
                        }
                    }
                }
                
                // '/exit' command
                case "/exit" -> {
                    // Halt continuation of program
                    bContinue = false;
                     
                    // Check if the client is connected to any server
                    if (socEndpoint.isClosed() == false) {
                        // Execute client disconnection protocol
                        try {
                            // Notify server that client will be disconnecting, and unregister client alias
                            dosOutput.writeUTF("L" + sClientAlias);
                            
                            // Close the connection
                            socEndpoint.close();
                            disInput = null;
                            dosOutput = null;
                            
                            // Reset client alias
                            sClientAlias = "";
                            
                            System.out.println("Disconnected from server at " + sServerAdd + ":" + nServerPort + ".");
                        } catch (IOException e) {
                            System.err.println("I/O ERROR: " + e.getMessage());
                        }
                    }
                }
                
                // '/get' command
                case "/get" -> {
                    // Check if (1) the client is connected to a server, (2) the client is registered to the server, (3) the user entered a file name
                    if (socEndpoint.isClosed()) {
                        System.out.println("ERROR: You are not connected to any File Exchange server.");
                    } else if (sClientAlias.isEmpty()) {
                        System.out.println("ERROR: You are not yet registered to this File Exchange server.");
                    } else if (scInput.hasNext() == false) {
                        System.out.println("ERROR: You have not entered any file name.\n"
                                + "Command format:\t/get [filename]");
                    } else {
                        // Get file name
                        String sFileName = scInput.next();
                        
                        // Send file name to server for download
                        try {
                            dosOutput.writeUTF("F" + sFileName);
                            
                            // Get server response, which is an integer indicating the size of the requested file (in bytes)
                            int nFileSize = disInput.readInt();
                            
                            // Evaluate server response
                            if (nFileSize == -1) {
                                System.out.println("ERROR: The requested file does not exist in the File Exchange server.");
                            } else {
                                // Create local file
                                FileOutputStream fosFile = new FileOutputStream(sFileName);
                                
                                // Receive file contents from server in a buffer and write it to local file
                                byte[] byFileBuffer = new byte[4096];
                                int bytesRead;
                                while (nFileSize > 0 && (bytesRead = disInput.read(byFileBuffer, 0, (int) Math.min(byFileBuffer.length, nFileSize))) != -1) {
                                    fosFile.write(byFileBuffer, 0, bytesRead);
                                    nFileSize -= bytesRead;
                                }
                                
                                System.out.println("File \"" + sFileName + "\" has been successfully downloaded.");
                            }
                        } catch (IOException e) {
                            System.err.println("I/O ERROR: " + e.getMessage());
                            sClientAlias = ""; // Reset client alias in the event of an error
                        }
                    }
                }
                
                // '/join' command
                case "/join" -> {
                    // Check if (1) the client is already connected to a server, (2) the user entered an IP address, and (3) the user entered a port number
                    if (socEndpoint.isClosed() == false) {
                        System.out.println("ERROR: You are already connected to a File Exchange server.");
                    } else if (scInput.hasNext() == false) {
                        System.out.println("ERROR: You have not entered any IP address.\n"
                                + "Command format:\t/join [ip_address] [port]");
                    } else if (scInput.hasNext() == false) {
                        System.out.println("ERROR: You have not entered any port number.\n"
                                + "Command format:\t/join [ip_address] [port]");
                    } else {
                        // Set server IP address
                        sServerAdd = scInput.next();

                        // Additional check if address is valid
                        if (isValidIPAddress(sServerAdd) == false) {
                            System.out.println("ERROR: You entered an invalid IPv4 address.");
                            sServerAdd = "";
                            
                        // Check if user entered a port number
                        } else if (scInput.hasNext() == false) {
                            System.out.println("ERROR: You have not entered any port number.\n"
                                    + "Command format:\t/join [ip_address] [port]");
                        } else {
                            // Set server port
                            // Try-catch system for port number validation
                            try {
                                // Attempt to parse the string as an integer
                                nServerPort = Integer.parseInt(scInput.next());
                            } catch (NumberFormatException e) {
                                System.out.println("ERROR: You entered an invalid server port number.");
                                nServerPort = -1;
                            }

                            // Connect to the server if all inputs are valid
                            if (nServerPort != -1) {
                                // Open (or reopen) client socket
                                socEndpoint = new Socket();
                                
                                // Connect socket to server, and if successful, connect the input and output streams too
                                if (connectSocket(socEndpoint, sServerAdd, nServerPort)) {
                                    try {
                                        disInput = new DataInputStream(socEndpoint.getInputStream());
                                        dosOutput = new DataOutputStream(socEndpoint.getOutputStream());
                                    } catch (IOException e) {
                                        System.err.println("I/O ERROR: " + e.getMessage());
                                    }
                                }
                            }
                        }
                    }
                }
                
                // '/leave' command
                case "/leave" -> {
                    // Check if the client is not connected to any server
                    if (socEndpoint.isClosed()) {
                        System.out.println("ERROR: You are not connected to any File Exchange server.");
                    } else {
                        try {
                            // Notify server that client will be disconnecting, and unregister client alias
                            dosOutput.writeUTF("L" + sClientAlias);
                            
                            // Close the connection
                            socEndpoint.close();
                            disInput = null;
                            dosOutput = null;
                            
                            // Reset client alias
                            sClientAlias = "";
                            
                            System.out.println("Disconnected from server at " + sServerAdd + ":" + nServerPort + ".");
                        } catch (IOException e) {
                            System.err.println("I/O ERROR: " + e.getMessage());
                        }
                    }
                }
                
                // '/register' command
                case "/register" -> {
                    // Check if (1) the client is connected to a server, (2) the client is already registered to the server, and (3) the user entered an alias
                    if (socEndpoint.isClosed()) {
                        System.out.println("ERROR: You are not connected to any File Exchange server.");
                    } else if (sClientAlias.isEmpty() == false) {
                        System.out.println("ERROR: You are already registered to this File Exchange server.");
                    } else if (scInput.hasNext() == false) {
                        System.out.println("ERROR: You have not entered any alias.\n"
                                + "Command format:\t/register [alias]");
                    } else {
                        // Get entered alias
                        sClientAlias = scInput.next();
                        
                        // Send alias to server for registration
                        try {
                            dosOutput.writeUTF("R" + sClientAlias);
                            
                            // Get server response
                            char cResponse = disInput.readChar();
                            
                            // Evaluate server response
                            switch (cResponse) {
                                case '/' -> System.out.println("You have successfully registered to the server with the alias " + sClientAlias + ".");
                                
                                case ':' -> {
                                    System.out.println("ERROR: The alias " + sClientAlias + " is already taken in this server.");
                                    sClientAlias = ""; // Reset client alias if it is taken
                                }
                                
                                case '*' -> {
                                    System.out.println("ERROR: An error occurred in the server.");
                                    sClientAlias = ""; // Reset client alias in the event of an error
                                }
                            }
                            
                        } catch (IOException e) {
                            System.err.println("ERROR: " + e.getMessage());
                            sClientAlias = ""; // Reset client alias in the event of an error
                        }
                    }
                }
                
                // '/rejoin' command
                case "/rejoin" -> {
                    // Check if (1) the client is already connected to a server, and (2) the client previously connected to a server
                    if (socEndpoint.isClosed() == false) {
                        System.out.println("ERROR: You are already connected to a File Exchange server.");
                    } else if (sServerAdd.isEmpty() || nServerPort == -1) {
                        System.out.println("ERROR: You have not connected to any File Exchange server during this session.");
                    } else {
                        // Reopen client socket
                        socEndpoint = new Socket();
                        
                        // Reconnect socket to previously-connected server, and if successful, reconnect the input and output streams too
                        if (connectSocket(socEndpoint, sServerAdd, nServerPort)) {
                            try {
                                disInput = new DataInputStream(socEndpoint.getInputStream());
                                dosOutput = new DataOutputStream(socEndpoint.getOutputStream());
                            } catch (IOException e) {
                                System.err.println("I/O ERROR: " + e.getMessage());
                            }
                        }
                    }
                }
                
                // '/store' command
                case "/store" -> {
                    // Check if (1) the client is connected to a server, (2) the client is registered to the server
                    if (socEndpoint.isClosed()) {
                        System.out.println("ERROR: You are not connected to any File Exchange server.");
                    } else if (sClientAlias.isEmpty()) {
                        System.out.println("ERROR: You are not yet registered to this File Exchange server.");
                    } else if (scInput.hasNext() == false) {
                        System.out.println("ERROR: You have not entered any file name.\n"
                                + "Command format:\t/store [filename]");
                    } else {
                        // Get file name
                        String sFileName = scInput.next();
                        
                        try {
                            // Find entered file in the local directory
                            File filToSend = new File(sFileName);
                            
                            // If file does not exist, throw an error, otherwise send to server
                            if (filToSend.exists() == false) {
                                System.out.println("ERROR: The file you entered does not exist.");
                            } else {
                                // First notify the server that a file will be sent, and send along the file name
                                dosOutput.writeUTF("S" + sFileName);
                                
                                // Get size of file and send to server
                                dosOutput.writeInt((int)filToSend.length());
                                
                                // After sending the size of the file, then send the actual contents of the file
                                // First convert the file into a FileInputStream object
                                try (FileInputStream fisFile = new FileInputStream(sFileName)) {
                                    // Set transfer buffer to 4 MB at a time
                                    byte[] byFileBuffer = new byte[4096];
                                    int bytesRead;

                                    // Read and send the contents of the file in a buffer
                                    while ((bytesRead = fisFile.read(byFileBuffer)) != -1) {
                                        dosOutput.write(byFileBuffer, 0, bytesRead);
                                    }
                                    
                                    System.out.println("File \"" + sFileName + "\" has been successfully uploaded to the server.");
                                    
                                } catch (IOException e) {
                                    System.err.println("I/O ERROR: " + e.getMessage());
                                }
                            }
                        } catch (IOException e) {
                            System.err.println("I/O ERROR: " + e.getMessage());
                        }
                    }
                }
                
                // For invalid commands
                default -> System.out.println("ERROR: \"" + sCommand + "\" is not recognized as a valid command.");
            }
        } while (bContinue);
        
        // Termination message
        System.out.println("Program terminated.\n");
    }
}

File: FileExchangeSystem_Client.py
# Machine Project Group 10
# ARGAMOSA, Daniel Cedric S.
# TAPIA, John Lorenzo N.

import socket
import threading
import json
import ipaddress
import os

# Setup client UDP socket
client_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

# buffer for files and messages
buffer = 4096 # buffer for files and messages
server_directory = os.path.abspath("ServerFolder")

# Determins whether the client is connected to the server
isConnected = False 
isRegistered = False 

server_address = None

def validIP(serverIP):
    try:
        ipaddress.ip_address(serverIP)
        return True
    except socket.error:
        return False

def validPort(nServerPort):
    try:
        port_num = int(nServerPort)
        return 0 < port_num < 65536  # Because the port range is 1-65535
    except ValueError:
        return False

def download(filename):
    global server_directory

    # Read the file from the server folder
    with open(os.path.join(server_directory, filename), 'rb') as server_file:
        file_content = server_file.read()

    # Write the file content into the client's 'Downloads' folder
    downloads_path = os.path.join(os.path.expanduser('~'), 'Downloads', filename)
    with open(downloads_path, 'wb') as client_file:
        client_file.write(file_content)


def upload(filename):
    global server_directory
    
    # Read the file from the directory
    with open(os.path.join(os.getcwd(), filename), 'rb') as upload_file:
        file_content = upload_file.read()

    # Write the file into the server folder
    with open(os.path.join(server_directory, filename), 'wb') as server_file:
        server_file.write(file_content)

# Processes commands sent to the server
def toServer(userInput):
    global isConnected
    global isRegistered
    global server_address
    
    if not userInput.startswith('/'):
        print(f"ERROR: {userInput} is not a valid command. Type /? for the list of available commands.")
        return
    
    input_list = userInput.split()
    command = input_list[0]
    params = input_list[1:]
      
    if command == "/join":
        if not isConnected:
            if len(params) != 2:
                print("ERROR: Invalid command syntax. Command format: /join [ip_address] [port]")
            else:
                try:
                    server_address = (params[0], int(params[1]))
                
                    # Send the "join" command to the server
                    print(f"Attempting to connect to the File Exchange Server at {server_address}")
                    client_socket.sendto(json.dumps({"command": "join"}).encode(), server_address)

                    client_socket.settimeout(3)

                    try:
                        acknowledgement = client_socket.recvfrom(buffer)
                        data = json.loads(acknowledgement[0].decode())
 
                        if data['command'] == "success":
                            print("Connection to the File Exchange Server is successful! Welcome!")
                            isConnected = True
                            client_socket.settimeout(None)

                    except socket.timeout:
                        print("ERROR: Connection attempt timed out. Please check the server availability and try again.")
                        server_address = None
                        return

                    except ConnectionResetError:
                        print("ERROR: Received unexpected data from the server. Please check the input paramaeters")
                        server_address = None
                        return

                    except Exception as e:
                        print(f"ERROR: {str(e)}")
                        server_address = None
                        return

                except Exception as e:
                    print(f"ERROR: Check IP Address and Server Port format. {str(e)}")
                    server_address = None
                    return    
        else:
            print('ERROR: You are already connected to a File Exchange server.')
      
    elif command == "/leave":
        if isConnected:
            if len(params) > 0:
                print("ERROR: Invalid command syntax. Command format: /leave")
            else:
                client_socket.sendto(json.dumps({"command":"leave"}).encode(), server_address)
                print(f"Disconnected from server at {server_address}. Thank you!")
                isConnected = False
                isRegistered = False 
                server_address = None
        else:
            print('ERROR: Please see if you are connected to a server first')
   
    elif command == "/register":
        if isConnected:
            if not isRegistered:
                if len(params) != 1:
                    print("ERROR: Invalid command syntax. Command format: /register [alias]")
                else:
                    client_socket.sendto(json.dumps({"command":"register", "alias":params[0]}).encode(), server_address)
                    
                    acknowledgement = client_socket.recvfrom(buffer)
                    data = json.loads(acknowledgement[0].decode())
                    message = data['message']
 
                    print(f">\n{message}") 

                    if data['command'] == "success":
                        isRegistered = True
            else:
                print('ERROR: You are already registered')
        else:
            print('ERROR: Please see if you are connected to a server first')
   
    elif command == "/dir":
        if isConnected:
            if isRegistered:
                if len(params) > 0:
                    print("ERROR: Invalid command syntax. Command format: /dir")
                else:
                    client_socket.sendto(json.dumps({"command":"dir"}).encode(), server_address)
            else:
                print('ERROR: Please see if you are a registered user first')
        else:
            print('ERROR: Please see if you are connected to a server first')
    
    elif command == "/get":
        if isConnected:
            if isRegistered:
                if len(params) != 1:
                    print("ERROR: Invalid command syntax. Command format: /get [filename.filetype]")
                else:

                    try:
                        client_socket.sendto(json.dumps({"command":"get", "filename":params[0]}).encode(), server_address)
                        # The code that uses the file functions in fromServer
                    except Exception as e:
                        return
            else:
                print('ERROR: Please see if you are a registered user first')
        else:
            print('ERROR: Please see if you are connected to a server first')

    elif command == "/store":
        if isConnected:
            if isRegistered:
                if len(params) != 1:
                    print("ERROR: Invalid command syntax. Command format: /store [filename.filetype]")
                else:
                    try:
                        client_socket.sendto(json.dumps({"command":"store", "filename":params[0]}).encode(), server_address)
                        # The code that uses the file functions in fromServer
                    except Exception as e:
                        return
            else:
                print('ERROR: Please see if you are a registered user first')
        else:
            print('ERROR: Please see if you are connected to a server first')
    
    elif command == "/all":
        if isConnected:
            if isRegistered:
                if len(params) == 0:
                    print("ERROR: Invalid command syntax. Command format: /all [message]")
                else:
                    message = ' '.join(params)
                    client_socket.sendto(json.dumps({"command":"all", "message":message}).encode(), server_address)
            else:
                print('ERROR: Please see if you are a registered user first')
        else:
            print('ERROR: Please see if you are connected to a server first')
   
    elif command == "/msg":
        if isConnected:
            if isRegistered:
                if len(params) < 2:
                    print("ERROR: Invalid command syntax. Command format: /msg [alias] [message]")
                else:
                    alias = params[0]
                    message = ' '.join(params[1:])
                    client_socket.sendto(json.dumps({"command":"msg", "alias":alias, "message":message}).encode(), server_address)
            else:
                print('ERROR: Please see if you are a registered user first')
        else:
           print('ERROR: Please see if you are connected to a server first')
    
    elif command == "/?":
        print("/?                           Display information about all valid commands")
        print("/all [message]               Send a message to every registered alias")
        print("/dir                         Display a directory of the files in the File Exchange Server")
        print("/get [filename.filetype]     Download a file of name [filename] from the File Exchange Server")
        print("/join [ip_address] [port]    Connects to the File Exchange Server")
        print("/leave                       Disconnects from the connected File Exchange Server")
        print("/msg [alias] [message]       Send a direct message to a single registered alias")
        print("/register [alias]            Register to the connected File Exchange server using a unique [alias]")
        print("/store [filename.filetype]   Upload a file of name [filename] to the File Exchange server - [filename] must exist inside the client's directory")
    
    else:
        print(f"ERROR: {userInput} is not a valid command. Type /? for the list of available commands.")

# Processes commands received from the server
def fromServer(data):
    command = data['command'] 
    
    # Acknowledge ping from server
    if command == 'ping':
        ping_ack = {'command': 'ping'}
        client_socket.sendto(json.dumps(ping_ack).encode(), server_address)
        return
    
    # Receive a global or error message 
    elif command == "all" or command == "error":
        message = data['message']
        
    # Receive a direct message
    elif command == "msg":
        alias = data['alias']
        message = f"[From {alias}]: {data['message']}"
    
    # Return success message
    elif command == "success":
        message = data['message']

        if 'filename' in data and 'type' in data:
            # Handles files
            if data['type'] == "store":
                upload(data['filename'])
            elif data['type'] == "get":
                download(data['filename'])
        
    print(f">\n{message}\n> ", end="")      

# Function thats being run in a thread
def receive_messages():
    global isConnected
    message = None
    
    while True:
        if isConnected:
            if isRegistered: # Majority of the features are locked behind a registered client
                try:
                    message = client_socket.recvfrom(buffer)
                    data = json.loads(message[0].decode())
                    fromServer(data)
                except ConnectionResetError:
                    print("ERROR: Connection to the File Exchange Server terminated. Disconnecting.")
                    isConnected = False
                except Exception as e:
                    print(f"ERROR Test: {str(e)}")

# Start a thread to continuously receive and print messages from the server
file_thread = threading.Thread(target = receive_messages)
file_thread.start()

print("------------------------------")
print("FILE EXCHANGE SYSTEM - CLIENT")
print("------------------------------")
print("Welcome!")
print("Please connect to a server to start...")
print("Enter a command. Type /? for get a list of all recognized commands")

while True:
    # Get a command from the user
    userInput = input("> ")
    toServer(userInput)

